platform/lf_nrf52_support.c

```diff
@@ -216,40 +221,28 @@ int lf_mutex_init(lf_mutex_t* mutex) {
  * @return 0 for success, or -1 if interrupted.
  */
 int lf_nanosleep(instant_t requested_time) {
+    uint32_t target_timer_val;
     instant_t target_time;
-    instant_t cur_time;
+
     lf_clock_gettime(&target_time);
     target_time += requested_time;
+    target_timer_val = (requested_time - _lf_time_epoch_offset) / 1000;
 
-    // printf("Entering nanosleep...\n");
-    INT_RAISED = 0;
-
-    // enable all interrupts
-    nrf_int* head;
-    head = int_list_head;
-    while (head != NULL) {
-        lf_mutex_unlock((lf_mutex_t*)head);
-        head = head->next;
-    }
-
-    // printf("At time " PRINTF_TIME " sleeping until " PRINTF_TIME "\n", cur_time, target_time);
-
-    do {
-        // cur_time gets initialized here
-        lf_clock_gettime(&cur_time);
-        if (INT_RAISED != 0) {
-            printf("DEBUG: Interrupt raised during lf_nanosleep.\n");
-            break;
-        }
-    } while(cur_time <= target_time);
-    // disable interrupts
-    head = int_list_head;
-    while (head != NULL) {
-        lf_mutex_lock((lf_mutex_t*)head);
-        head = head->next;
-    }
-    int result = (INT_RAISED == 0)? 0 : -1;
-    INT_RAISED = 0;
+    // enable nvic
+    sd_nvic_critical_region_exit(_lf_nested_region);
+    
+    // reset by timer interrupt callback
+    _lf_interrupted = true;
```

Comment:
Possible race condition here. I think `_lf_interrupted` needs to be set before exiting the critical section. Otherwise, an interrupt could occur between line 232 and line 235 and `_lf_interrupted` will have the wrong value.

Also, `_lf_interrupted` is misnamed because it has value `true` when an interrupt has *not* occurred.

Reply:
moved and added some comments. Its a little confusing but _lf_interrupted is default set true meaning that an interrupt other than the requested time interrupt from the timer fired and exited the sd_app_evt_wait(). The timer callback sets _lf_interrupted false meaning no need to re check the event queue.

Reply:
I think setting to `false` by default makes more sense.